### ISSUE TITLE
Rain - something wrong with the data received

### ISSUE THREAD

**Issue by LloydR** [lheijst/weewx-rtld#15]
This is what I found - 
        1. data: 03:57:08.612942 E003BE730300E26A 34901
        2. data: 03:57:08.613204 E0019E310300E26A 34902
        3. data: 03:57:29.113525 E001BE730300A6E9 34909
notice there are 2 ea data strings at 03:57:08 within milliseconds of each other. The 3rd byte (counting from 0 and from the left to right) is rain clicks and goes from 73 to 31  and back to 73 causing a report of 1.3" of rain.  This happened twice once at 01:13;29 and 03:57:08. I expect that is probably the cause of not only rain but wind and even temperature. There should not be data coming from the GO program that quickly, neither should there be duplicate data such as
DEBUG user.rtldavis: data: 12:07:33.906445 8003BC166B00A202 46260 0 0 0 0 msg.ID=0
DEBUG user.rtldavis: info: 12:07:33.908690 duplicate packet: 8003BC166B00A202
One would think the CRC would kick in but the packets have good CRC.

Further investigation warranted 

11/7/22  Not only wind (I have noticed that also) but also temperature and probably any value coming into weewx-rtldavis

weewx[15687] DEBUG user.rtldavis: data: 05:00:31.352998 8004770E6A00B19E 163 0 0 0 0 msg.ID=0
weewx[15687] DEBUG user.rtldavis: data: 05:00:31.353372 800433062A00B19E 164 0 0 0 0 msg.ID=0
weewx[15687] DEBUG user.rtldavis: data: 05:00:41.601651 8002770E6B004F2A 168 0 0 0 0 msg.ID=0

11/14/22 Got an anomalous wind reading - wind was zero then all of a sudden weewx-rtldaavis reports 27 mph. Two things - once again CRC OK, Type 03 which is unknown and should have never been processed.

weewx[15687] DEBUG user.rtldavis: data_pkt: {'channel': 1, 'bat_iss': 0, 'wind_speed_ec': 0, 'wind_speed_raw': 0, 'wind_dir': 122.5494071146245, 'wind_speed': 0.0, 'temperature': -13.25, 'curr_cnt0': 111000, 'curr_cnt1': 0, 'curr_cnt2': 0, 'curr_cnt3': 0}
weewx[15687] DEBUG user.rtldavis: data: 03:18:09.714547 3019B480A320EF9F 111001 0 0 0 0 msg.ID=0
weewx[15687] DEBUG user.rtldavis: data_pkt: {'channel': 1, 'bat_iss': 0, 'wind_speed_ec': 27, 'wind_speed_raw': 25, 'wind_dir': 250.9683794466403, 'wind_speed': 12.07005, 'curr_cnt0': 111001, 'curr_cnt1': 0, 'curr_cnt2': 0, 'curr_cnt3': 0}

11/29/22 Rain again - 3rd byte (starting from 0 and left) goes 5C to 1C back to 5C within milliseconds
weewx[14677] DEBUG user.rtldavis: data: 07:52:59.119600 E0015B5C030034B1 51660 0 0 0 0 msg.ID=0
weewx[14677] DEBUG user.rtldavis: data: 07:53:09.368426 E0025B5C0300DA63 51664 0 0 0 0 msg.ID=0
weewx[14677] DEBUG user.rtldavis: data: 07:53:09.370324 E0021B1C02009A63 51665 0 0 0 0 msg.ID=0
weewx[14677] DEBUG user.rtldavis: data: 07:53:19.619408 E0015B5C030034B1 51669 0 0 0 0 msg.ID=0

12-21-22 Wind again - this time the type is 0
306776 weewx[14677] DEBUG user.rtldavis: data: 12:47:58.826613 00F29BE9A3EB999D 12499 0 0 0 0 msg.ID=0
 306777 
 306778 weewx[14677] DEBUG user.rtldavis: wind_speed_raw=0f2 wind_dir_raw=0x09b
 306779 weewx[14677] DEBUG user.rtldavis: WS=108.18341111111111 WD=217.17391304347825 WS_raw=242 WS_ec=242 WD_raw=155 WD_pro=217.173913
 306779 04347825 WD_vue=218.26875
 306780 weewx[14677] ERROR user.rtldavis: unknown message type 0x0

 429318 weewx[14677] DEBUG user.rtldavis: data: 21:39:52.820657 003B264537339CA9 12201 0 0 0 0 msg.ID=0
 429319 
 429320 weewx[14677] DEBUG user.rtldavis: wind_speed_raw=03b wind_dir_raw=0x026
 429321 weewx[14677] DEBUG user.rtldavis: rx0=36, rx1=40, ry0=55, ry1=60, x0=0, x1=0, y0=0, y1=0, x=38, y=59
 429322 weewx[14677] DEBUG user.rtldavis: WS=26.375294444444442 WD=59.015810276679844 WS_raw=59 WS_ec=59 WD_raw=38 WD_pro=59.0158102766
 429322 79844 WD_vue=53.7375
 429323 weewx[14677] ERROR user.rtldavis: unknown message type 0x0

1/31/23
I changed the golang program to reject any reading , per transmitter, that comes in sooner than 2 seconds (scan time is 2.5 seconds).  Finally got 2 temperature readings, both passed CRC, one was -7.5 and the very next one was -10.7. Continuing to run with my changes in on wind and "duplicates" so I can somewhat confidently say this is a good patch until someone figures out why the SDR program does send garbage.
    1. weewx[28599] DEBUG user.rtldavis: data: 18:17:19.850988 800573FB4B00EC4B 8329 0 0 0 0 msg.ID=0
    2. weewx[28599] DEBUG user.rtldavis: info: 18:17:19.851412 time between reads < 2 seconds, skipping 800573F94900E449

2/4/23
weewx[28599] DEBUG user.rtldavis: data: 18:09:38.215962 A005987A29002294 4602 0 0 0 0 msg.ID=0
weewx[28599] DEBUG user.rtldavis: info: 18:09:39.886586 time between reads < 2 seconds, skipping E0EDE647FC791312

**Comment by guidocioni** [comment]
I noticed this as well :(

**Comment by gary-hammer** [comment]
For me, it's wind
I've gotten wind and/or gusts of up to 125 MPH

**Comment by guidocioni** [comment]
> For me, it's wind I've gotten wind and/or gusts of up to 125 MPH

Yep, I have spike outliers in both rain and winds. They are getting so frequent that it's hard to put the station reliably online :( 
Haven't found a way to correct them but I'm sure it's not about poor reception.

**Comment by LloydR** [comment]
At least I am not the only one. A interim fix would be to look at the time stamp. If less than say 1/2 second then throw the value away. However I have not looked at how to do that with multiple stations. There must be some way to distinguish between stations. I know Luc has the "duplicate values" set up with a previous reading and then current reading and if same == duplicate.  Time stamp may be a little more difficult depending on how the multiple stations are handled.

**Comment by guidocioni** [comment]
> At least I am not the only one. A interim fix would be to look at the time stamp. If less than say 1/2 second then throw the value away. However I have not looked at how to do that with multiple stations. There must be some way to distinguish between stations. I know Luc has the "duplicate values" set up with a previous reading and then current reading and if same == duplicate. Time stamp may be a little more difficult depending on how the multiple stations are handled.

Are you sure the problem with these spikes is only related to the observations being really close in time?

**Comment by LloydR** [comment]
In reply to guidocioni - Am I sure the problem is packets being close in time.  I can't guarantee it but given 2 separate data points (Rain & Temp) it sure seems to me that is the problem. I am in the country, no other Davis ISS around me, I only have one Davis ISS and I am getting duplicate packages - There is something wrong with the Golang program or the RTL-SDR library.  The data is supposed to be coming in every 2.5 seconds. Anything other than that is bad data in my case. By the way the duplicate packages and the data shown in the statement of the problem are from the Golang debug. Trying to get my head around what the Golang program does has proven to be difficult for me.

**Comment by guidocioni** [comment]
> In reply to guidocioni - Am I sure the problem is packets being close in time. I can't guarantee it but given 2 separate data points (Rain & Temp) it sure seems to me that is the problem. I am in the country, no other Davis ISS around me, I only have one Davis ISS and I am getting duplicate packages - There is something wrong with the Golang program or the RTL-SDR library. The data is supposed to be coming in every 2.5 seconds. Anything other than that is bad data in my case. By the way the duplicate packages and the data shown in the statement of the problem are from the Golang debug. Trying to get my head around what the Golang program does has proven to be difficult for me.

Yeah I also don't know golang, I can only comment on the Python part. But, in theory, if the problem is with these milliseconds package, we could just put a `sleep` command or some kind of guard in the code to prevent these duplicated packages to be received.

BTW I also have a similar setup with just one station and no interference.

**Comment by gary-hammer** [comment]
I find it notable that all of us have different observations that are affected.
For me, it is wind and only wind. I have limited the QC check values in WeeWX and that helps as I'm not expecting more than 85 MPH winds where I live.
Still, overnight I have a 64 MPH gust recorded. Then back to the 0-3 MPH gentle breeze.

I too have eliminated all the external sources of interference and the remaining suspect is the driver.

**Comment by LloydR** [comment]
Looking at Golang code in rtldavis I have not figured out how the bad data and duplicate gets in but it is probably because of the goroutine (go func() {   line 270) . 

In debugging it appears that any log data that has the timestamp come from Golang rtldavis and not weewx-rtldavis. (DEBUG user.rtldavis: info: 18:00:01.149015).  

Because Golang rtldavis tries to handle multiple sending devices it can not just check on current time vs previous time. It has to be indexed by transmitter number.

**Comment by LloydR** [comment]
Looks like wind is somewhat different - the weewx-rtldavis looks like it processes wind even if there is a unkown packet and the packet gets through the CRC check.  That seems to be a major problem with this software suite since there are bogus packets that get through. Maybe it should check for a good packet before processing the wind data. Also noticed that the wind gust values Type 0x9 are not used; which is what is recommended since all it is is the highest value of the wind measurement for the last 9 packets.

**Comment by WeatheredScientist** [comment]
> Disclosure: the analysis and the patch were done by Claude (Anthropic), working with me on my station.
> I have read it all and I stand behind it, but the credit for the digging is not mine.

I think this one is solved, and the answer was sitting in @LloydR's packet dumps at the top of the
thread the whole time. There are two separate bugs stacked on each other, which is probably why it has
been so slippery.

**First: the receiver is decoding some transmissions twice.**

@LloydR's first two frames arrive **262 µs apart**. The ISS only transmits every ~2.5 s, so that cannot
be two transmissions — it is one RF burst being decoded twice, with the second copy picking up bit
errors on the way.

I went looking for that on my own station and it is unambiguous: **61 frames arriving 1.4–10 ms (median
2.0 ms) after a byte-identical frame from the same transmitter, in two hours.** Call it 700 a day. An
ISS cannot transmit twice 2 ms apart, so the receiver is making the copy.

Nearly all of those copies fail CRC and get dropped quietly inside the Go binary, which is why nobody
sees them. But **CRC-16 only guarantees it catches *single*-bit errors.** A multi-bit pattern can land
on a valid codeword and sail straight through — and that is exactly what LloydR's two frames are: they
differ in **4 bits** and **both pass CRC**. I ran them through `weewx.crc16` to be sure.

So the checksum is not protecting us here, and — @guidocioni, you called this — **it is not a reception
problem.** Better antennas and better SNR will not fix it.

It also explains why the existing `data != self._last_pkt` guard misses these: a *corrupted* near-copy
is not equal to the previous packet, so only the harmless identical case gets caught.

**Second: the driver turns one bad packet into inches of rain.**

This is the part that actually hurts, and it is pure software. `_data_to_packet()` treats **any** negative
counter delta as a 127→0 wraparound and adds 128:

```python
if rain_count < 0:
    loginf("rain counter wraparound detected rain_count=%s" % rain_count)
    rain_count += 128
```

That is correct for a real wraparound. But a corrupt reading can produce a *small* negative delta, and
then adding 128 invents rain out of nothing.

Run @LloydR's own counter values through it (115 → 49 → 115):

- `49 - 115 = -66` → `-66 + 128` = **62 tips = 0.62"**
- `115 - 49 = +66` → unbounded → **66 tips = 0.66"**

**1.28" total** — which is the ~1.3" he reported. The second reading was just the sensor coming back to
the truth, and the driver charged him for it.

I have had the identical signature twice on my station, 1.28" both times, on days with **no rain at
all** — once with `rain counter wraparound detected rain_count=-64` right there in the log. Both checked
against a co-located WeatherLink Live console on the same ISS, which recorded nothing.

**The fix**

I have opened #22. The core of it:

```python
MAX_PLAUSIBLE_TIPS = 60           # tips between two received packets
RAIN_WRAPAROUND_THRESHOLD = -100  # a genuine wraparound is about -127

def rain_delta_tips(last_count, new_count, max_tips=MAX_PLAUSIBLE_TIPS):
    if last_count is None:
        return 0
    delta = new_count - last_count
    if delta < RAIN_WRAPAROUND_THRESHOLD:   # genuine 127 -> 0 wraparound
        delta += 128
    if delta < 0 or delta > max_tips:       # glitch, or implausible spike
        return None                         # missing, not zero
    return delta
```

Only a *large* negative delta counts as a wraparound. An implausibly large positive delta is rejected
too, because the high bit can also flip *on* directly. And a rejected reading returns `None`, so weewx
records it as **missing** rather than as rain — an honest gap beats fabricated rainfall.

60 tips is 0.60" between two received packets. The world one-minute rainfall record is about 1.23", so
that is generous headroom, while still sitting well below the characteristic 64/128-tip glitches.
Genuine wraparounds are untouched: `127 → 0` still gives 1 tip. It has been running on my station since
May.

**Two things I want to be straight about**

1. **This fixes the rain counter, not `rainRate` — and the rate is a different bug entirely, in the ISS,
   not in the driver.** I chased my own phantom rate events and the data is unambiguous: the rate held
   for **exactly the ISS's ~15-minute rain-rate timeout**, the implied tip interval stayed in a tight
   8.5–10 s band the whole time, and **the tip counter never moved** (zero rain in every one of those
   records). Reaching those raw values from the `0x3FF` "no rain" sentinel needs ~6 bit-flips **in every
   packet for 16 minutes** — RF corruption cannot do that. The ISS was genuinely sending it.

   Both my events happened overnight at **94% humidity, a 1.7 °F dewpoint spread, and dead calm.** Best
   explanation: **condensation trips the reed switch often enough to start the rate timer, but never
   enough water accumulates to actually tip the bucket.** Rate register set, counter untouched — which is
   exactly what the data shows.

   So: don't look for the rate bug in the driver. No decode-layer fix will touch it.

2. **@gary-hammer, your 64 mph gusts are almost certainly the same bug in a different field.** Same
   mechanism — a CRC-valid corrupted duplicate — landing in the wind bytes instead of the rain counter.
   A plausibility check at decode time is the only defense available at that layer, since CRC has
   already been fooled. I run one on temperature, humidity, wind, UV and radiation now. Happy to send it
   as a separate PR if there is appetite; I deliberately kept #22 to the rain fix so it stays reviewable.

@LloydR — your packet dumps are the entire reason this was findable. Thank you for posting them, and for
leaving them up for three years. And if I have misread the wraparound logic, I would genuinely rather be
told than be right.
